### PR TITLE
Adds useStateSafely hook

### PR DESCRIPTION
**Pull request by jkilzi**
This new hook is a drop-in replacement for `React.useState`. 

Should be used when:
- The internal state of a component is altered by delegating the operation to a nested component.
- The internal state changes inside an async operation.

In general, this hook prevents changing the state of an unmounted component.

### CODE REVIEW

File: src/common/hooks/useStateSafely.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import type { Dispatch, SetStateAction } from 'react';
+import { useCallback, useEffect, useRef, useState } from 'react';
+
+export default function useStateSafely<S>(initialState: S): [S, Dispatch<SetStateAction<S>>] {
```
Comment (rawagner) [inline src/common/hooks/useStateSafely.ts:4]: We usually use arrow functions, why mixing the style with `function` keyword ?

File: src/common/components/hosts/DiskRole.tsx [hunk 2]
```diff
@@ -81,9 +82,9 @@ const DiskRoleDropdown: React.FC<DiskRoleDropdownProps> = ({
         // TODO(mlibra): Improve for the case onDiskRole === undefined
         setOpen(false);
       };
-      asyncFunc();
+      void asyncFunc();
```
Comment (rawagner) [inline src/common/components/hosts/DiskRole.tsx:85]: what is the reason for adding `void` here ?

File: src/common/hooks/useStateSafely.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import type { Dispatch, SetStateAction } from 'react';
+import { useCallback, useEffect, useRef, useState } from 'react';
```
Comment (rawagner) [inline src/common/hooks/useStateSafely.ts:2]: why do we have separate imports for functions and types ?

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: > This operator allows evaluating expressions that produce a value into places where an expression that evaluates to undefined is desired.
Since we ignore the returned promise, IMHO it makes sense to use it. The code gets executed as usual, the promise still fulfills or rejects anyway.

File: src/common/hooks/useStateSafely.ts (hunk 3 above)
Comment (jkilzi) [reply src/common/hooks/useStateSafely.ts:2]: A: in order to make it clear what types the module is consuming and which functions it uses from the React package.

File: src/common/hooks/useStateSafely.ts (hunk 1 above)
Comment (jkilzi) [reply src/common/hooks/useStateSafely.ts:4]: A: This module exports a single default function. Using a lambda here (as in many other places in our codebase) would demand declaring it in one statement and then adding an additional statement for the exported member.

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (mareklibra) [inline src/common/components/hosts/DiskRole.tsx:85]: why is this needed?

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: https://github.com/openshift-assisted/assisted-ui-lib/pull/936#discussion_r740035500

Comment (mareklibra) [reply src/common/components/hosts/DiskRole.tsx:85]: I missed the response above. IMHO, it brings noise in the code.
But I would be ok with merging it.

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: You can think of it as an explicit way of saying: "just call this (async) function, I (explicitly) don't care about the outcome"

File: src/common/hooks/useStateSafely.ts (hunk 3 above)
Comment (mareklibra) [reply src/common/hooks/useStateSafely.ts:2]: IMHO, it would make sense when applied consistently throughout the code, which is not recent state.

Comment (jkilzi) [reply src/common/hooks/useStateSafely.ts:2]: IMHO, it's a good chance to start adopting it.

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: so without `void` it means something different than `"just call this (async) function, I (explicitly) don't care about the outcome"` ?

File: src/common/hooks/useStateSafely.ts (hunk 3 above)
Comment (rawagner) [reply src/common/hooks/useStateSafely.ts:2]: I disagree with adopting this, unless we have automatic prettier/eslint rule for it

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: If you and I have the same dev environment setup, you should see an annoying linting message saying that you're ignoring the value returned by the promise and then the alternative will be to silent the linter for this line or just use the void operator as I did.

File: src/common/hooks/useStateSafely.ts (hunk 1 above)
Comment (rawagner) [reply src/common/hooks/useStateSafely.ts:4]: Ok, Im fine with it

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: linter setup is defined by `.eslintrc` in root folder, isnt it ? Why do you use different setup ? The linter validation should be the same as CLI config otherwise we start adding random stuff (like this `void`) and later somebody can remove it because our CLI linter config changes and it starts to complain etc. Endless loop.

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: I totally agree, with you. My setup consumes the eslint configuration from `.eslintrc` located at the root folder. I'm wondering whether you see this message about the returned promise being ignored, as I do, on this line or not?

Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: I dont.
Even running `yarn test:lint` does not show any warnings

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: image

Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: Is this really an eslint rule ? I could not find anything like that.
Isnt it just some smartness build in IntelliJ ?

File: src/common/hooks/useStateSafely.ts (hunk 3 above)
Comment (jkilzi) [reply src/common/hooks/useStateSafely.ts:2]: Can anybody mention why this is a bad practice? So far this discussion is only a matter of "taste".
- Line 1 gets completely erased after the code is transpiled.
- It is not harming the project in any form.
- It helps readers differentiate between types and functions exported by the package.

*Honestly, this is not mandatory. I don't expect you to start following this convention if you don't like it.*

Anyway, [this rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-imports.md) should help (despite I don't think it should be using it).

File: src/common/components/hosts/DiskRole.tsx (hunk 2 above)
Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: Yep seems it's coming from inteliJ. Is that a bad thing?

Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: I guess its fine. I was mostly worried that you have different eslint config. 
If you dont mind putting void on other places in the lib in the future (as people who do not use InteliJ wont do it), instead of turning that inspection info off, then Im good :)

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: > If you don't mind putting void on other places in the lib in the future

I'll even do that with pleasure :)

Comment (mareklibra) [reply src/common/components/hosts/DiskRole.tsx:85]: I still think it's just about IDE setup. Such syntactic sugar just makes the code/language talkative, harder to read.

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: Well, this is JavaScript @mareklibra. May I ask what part exactly do you find hard to read, maybe I can explain it to you?

Comment (mareklibra) [reply src/common/components/hosts/DiskRole.tsx:85]: My point is not in that. I was just referring to style - to avoid whatever is not needed.

Comment (batzionb) [reply src/common/components/hosts/DiskRole.tsx:85]: [There's an interesting discussion in github about this ](https://stackoverflow.com/questions/41278900/intellij-idea-warning-promise-returned-is-ignored-with-aysnc-await)
I'm assuming Joni saw this...
The solution Joni mentioned is one of the answers.
I thought the solution of adding .catch might also be a good option, since error handling is always a good idea...

Comment (rawagner) [reply src/common/components/hosts/DiskRole.tsx:85]: `try catch` makes more sense to me than adding `void`.

Comment (jkilzi) [reply src/common/components/hosts/DiskRole.tsx:85]: @rawagner you cannot use a try-catch there, `asyncFunc` is not being called inside an async block.
What you can do is actually something like this: `asyncFunc().catch(/*...*/);`
